keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1117 @@
+
+## Design Details
+
+### DRA Scheduler Plugin Design Overview
+
+This document outlines the design of the DRA Scheduler Plugin, focusing on the handling of fabric devices.
+Key additions include new attributes for device identification, enhancements to `AllocatedDeviceStatus`, and the process for handling `ResourceSlices` upon attachment failure.
+The composable controller design is also discussed, emphasizing efficient utilization of fabric devices.
+
+![proposal](proposal.jpg)
+
+#### BasicDevice Enhancements
+
+To indicate whether a device is a fabric device, fields are added to the `Basic` within `Device`.
+These fields will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
+
+```go
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+type Device struct {
+    // Name is unique identifier among all devices managed by
+    // the driver in the pool. It must be a DNS label.
+    //
+    // +required
+    Name string
+
+    // Basic defines one device instance.
+    //
+    // +optional
+    // +oneOf=deviceType
+    Basic *BasicDevice
+}
+
+// BasicDevice represents a basic device instance.
+type BasicDevice struct {
+    // Attributes defines the set of attributes for this device.
+    // The name of each attribute must be unique in that set.
+    //
+    // The maximum number of attributes and capacities combined is 32.
+    //
+    // +optional
+    Attributes map[QualifiedName]DeviceAttribute
+
+    // BindingGates defines the gates for binding.
+    //
+    // +optional
+    BindingGates []string
+
+    // BindingFailureGates defines the gates for binding failure.
+    //
+    // +optional
+    BindingFailureGates []string
+
+    // UsageRestrictedToNode indicates if the usage of an allocation involving this device
+    // has to be limited to exactly the node that was chosen when allocating the claim.
+    //
+    // +optional
+    UsageRestrictedToNode bool
+
+    // BindingTimeout indicates the prepare timeout period(minute).
+    // If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
+    //
+    // +optional
+    BindingTimeout int
+}
+```
+
+#### AllocatedDeviceStatus Enhancements
+
+The `BindingGates` and `BindingFailureGates` fields within `AllocatedDeviceStatus` are used to indicate the status of the device attachment.
+These fields will contain a list of conditions, each representing a specific state or event related to the device.
+
+For this feature, following fields are added:
+
+```go
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+    ...
+    // BindingGates defines the gates for binding.
+    //
+    // +optional
+    BindingGates map[string]bool
```

Comment:
I thought the idea was to use the existing `AllocatedDeviceStatus.Conditions` field.

Reply:
This is here to provide the "as-was" list of gates (the list of gates as they were defined in the slice at the time of the allocation) - I think it should just be a `[]string`

Reply:
Okay, now I get it. More comments on the field definitely would have helped.

Ack on `[]string`.